rust-runtime/aws-smithy-compression/src/gzip.rs

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::{Compress, CompressionOptions};
+use aws_smithy_runtime_api::box_error::BoxError;
+use flate2::write::GzEncoder;
+use std::io::prelude::*;
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub(crate) struct Gzip {
+    compression: flate2::Compression,
+}
+
+impl Gzip {
+    fn compress_bytes(&self, bytes: &[u8], writer: impl Write) -> Result<(), BoxError> {
+        let mut encoder = GzEncoder::new(writer, self.compression);
+        encoder.write_all(bytes)?;
+        encoder.try_finish()?;
+
+        Ok(())
+    }
+}
+
+impl Compress for Gzip {
+    fn compress_bytes(&mut self, bytes: &[u8], writer: &mut dyn Write) -> Result<(), BoxError> {
+        Gzip::compress_bytes(self, bytes, writer).map_err(Into::into)
```

Comment:
Is the `flate2` gzip implementation deterministic by default? Was reading and it looks like the GNU version is not unless you use `-n` since it includes a timestamp in the file. Note sure if that really matters for our use-case.

Reply:
The test was broken b/c I mounted the interceptor twice in codegen and I was doing snapshot testing. Once I removed the extra interceptor things worked fine. As far as I can tell, GZIP is deterministic in CI b/c the tests looking for a specific output are passing. If GZIP wasn't deterministic then we'd have to use a decoder to unzip and then check the contents. I think I do that in the compression crate somewhere.

I'm going to ignore it until it actually fails. LMK if you can think of a good test to add.